pkg/storage/tsdb/bucketindex/loader.go

```diff
@@ -0,0 +1,253 @@
+	"github.com/cortexproject/cortex/pkg/util/services"
+)
+
+const (
+	// readIndexTimeout is the maximum allowed time when reading a single bucket index
+	// from the storage. It's hard-coded to a reasonably high value.
+	readIndexTimeout = 15 * time.Second
+)
+
+type LoaderConfig struct {
+	CheckInterval         time.Duration
+	UpdateOnStaleInterval time.Duration
+	UpdateOnErrorInterval time.Duration
+	IdleTimeout           time.Duration
+}
+
+// Loader is responsible to lazy load bucket indexes and, once loaded for the first time,
+// keep them updated in background. Loaded indexes are automatically offloaded once the
+// idle timeout expires.
+type Loader struct {
+	services.Service
+
+	bkt    objstore.Bucket
+	logger log.Logger
+	cfg    LoaderConfig
+
+	indexesMx sync.RWMutex
+	indexes   map[string]*cachedIndex
+
+	// Metrics.
+	loadAttempts prometheus.Counter
+	loadFailures prometheus.Counter
+	loadDuration prometheus.Histogram
+	loaded       prometheus.GaugeFunc
+}
+
+// NewLoader makes a new Loader.
+func NewLoader(cfg LoaderConfig, bucketClient objstore.Bucket, logger log.Logger, reg prometheus.Registerer) *Loader {
+	l := &Loader{
+		bkt:     bucketClient,
+		logger:  logger,
+		cfg:     cfg,
+		indexes: map[string]*cachedIndex{},
+
+		loadAttempts: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cortex_bucket_index_loads_total",
+			Help: "Total number of bucket index loading attempts.",
+		}),
+		loadFailures: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cortex_bucket_index_load_failures_total",
+			Help: "Total number of bucket index loading failures.",
+		}),
+		loadDuration: promauto.With(reg).NewHistogram(prometheus.HistogramOpts{
+			Name:    "cortex_bucket_index_load_duration_seconds",
+			Help:    "Duration of the a single bucket index loading operation in seconds.",
+			Buckets: []float64{0.01, 0.02, 0.05, 0.1, 0.2, 0.3, 1, 10},
+		}),
+	}
+
+	l.loaded = promauto.With(reg).NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "cortex_bucket_index_loaded",
+		Help: "Number of bucket indexes currently loaded in-memory.",
+	}, l.countLoadedIndexesMetric)
+
+	// Apply a jitter to the sync frequency in order to increase the probability
+	// of hitting the shared cache (if any).
+	checkInterval := util.DurationWithJitter(cfg.CheckInterval, 0.2)
+	l.Service = services.NewTimerService(checkInterval, nil, l.updateCachedIndexes, nil)
+
+	return l
+}
+
+// GetIndex returns the bucket index for the given user. It returns the in-memory cached
+// index if available, or load it from the bucket otherwise.
+func (l *Loader) GetIndex(ctx context.Context, userID string) (*Index, error) {
+	l.indexesMx.RLock()
+	if entry := l.indexes[userID]; entry != nil {
+		idx := entry.index
+		err := entry.err
+		l.indexesMx.RUnlock()
+
+		// We don't check if the index is stale because it's the responsibility
+		// of the background job to keep it updated.
+		entry.requestedAt.Store(time.Now().Unix())
```

Comment:
I'd suggest making update of requestedAt timestamp a method on `*cachedIndex`, to keep the logic at single place.